plugins/module_utils/ansible_freeipa_module.py

```diff
@@ -309,3 +349,159 @@ def is_ipv6_addr(ipaddr):
+        pass
+
+    def api_command(self, command, name=None, args=None):
+        if args is None:
+            args = {}
+
+        if name is None:
+            return api_command_no_name(self, command, args)
+
+        return api_command(self, command, name, args)
+
+    def __enter__(self):
+        principal = self.ipa_params.ipaadmin_principal
+        password = self.ipa_params.ipaadmin_password
+
+        try:
+            if not valid_creds(self, principal):
+                self.ccache_dir, self.ccache_name = temp_kinit(
+                    principal, password,
+                )
+
+            api_connect()
+
+        except Exception as excpt:
+            self.fail_json(msg=str(excpt))
+        else:
+            self.ipa_connected = True
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_val:
+            self.fail_json(msg=str(exc_val))
+
+        # TODO: shouldn't we also disconnect from api backend?
+        temp_kdestroy(self.ccache_dir, self.ccache_name)
+
+        self.exit_json(changed=self.changed, user=self.exit_args)
+
+    def get_command_errors(self, command, result):
+        # Get all errors
+        # All "already a member" and "not a member" failures in the
+        # result are ignored. All others are reported.
+        errors = []
+        for item in result.get("failed", tuple()):
+            failed_item = result["failed"][item]
+            for member_type in failed_item:
+                for member, failure in failed_item[member_type]:
+                    if (
+                        "already a member" in failure
+                        or "not a member" in failure
+                    ):
+                        continue
+                    errors.append(
+                        "%s: %s %s: %s"
+                        % (command, member_type, member, failure)
+                    )
+
+        if len(errors) > 0:
+            self.fail_json(", ".join("errors"))
+
+    def add_ipa_command(self, command, name=None, args=None):
+        self.ipa_commands.append((name, command, args or {}))
+
+    def _run_ipa_commands(self):
+        result = None
+
+        for name, command, args in self.ipa_commands:
+            try:
+                result = self.api_command(command, name, args)
+            except Exception as excpt:
+                self.fail_json(msg="%s: %s: %s" % (command, name, str(excpt)))
+            else:
+                if "completed" in result:
+                    if result["completed"] > 0:
+                        self.changed = True
+                else:
+                    self.changed = True
+
+            self.get_command_errors(command, result)
+
+    def require_ipa_attrs_change(self, command_args, ipa_attrs):
+        equal = compare_args_ipa(self, command_args, ipa_attrs)
+        return not equal
+
+    def plog(self, value):
+        self.log(pformat(value))
+
+    def pdebug(self, value):
+        self.debug(pformat(value))
```

Comment:
It should also provide warn and error wrappers.

Reply:
I was in doubt even about keeping the "plog". The reason is that pformat will result the logs to be multi-line, which is nice for human readying but a little problematic for parsing. I haven't added to all logging levels to not encourage module developers to use pformat.

What do you think about removing plog and only keeping pdebug?

Reply:
That would be nice, as debug is (usually) geared towards humans, and the other levels for logs that might be parsed.